core/common/src/main/java/alluxio/underfs/ObjectLowLevelOutputStream.java

```diff
@@ -0,0 +1,392 @@
+ * <p>
+ * In flush(), we upload the buffered data if they are bigger than 5MB
+ * and wait for all uploads to finish. The temp files will be deleted after uploading successfully.
+ * <p>
+ * In close(), we upload the last part of data (if exists), wait for all uploads to finish,
+ * and complete the multipart upload.
+ * <p>
+ * close() will not be retried, but all the multipart upload
+ * related operations(init, upload, complete, and abort) will be retried.
+ * <p>
+ * If an error occurs and we have no way to recover, we abort the multipart uploads.
+ * Some multipart uploads may not be completed/aborted in normal ways and need periodical cleanup
+ * by enabling the {@link PropertyKey#UNDERFS_CLEANUP_ENABLED}.
+ * When a leader master starts or a cleanup interval is reached, all the multipart uploads
+ * older than clean age will be cleaned.
+ */
+@NotThreadSafe
+public abstract class ObjectLowLevelOutputStream extends OutputStream {
+  protected static final Logger LOG = LoggerFactory.getLogger(ObjectLowLevelOutputStream.class);
+
+  protected final List<String> mTmpDirs;
+
+  /**
+   * Only parts bigger than 5MB could be uploaded through multipart upload,
+   * except the last part.
+   */
+  protected static final long UPLOAD_THRESHOLD = 5L * Constants.MB;
+
+  /** Bucket name of the object storage bucket. */
+  protected final String mBucketName;
+
+  /** Key of the file when it is uploaded to object storage. */
+  protected final String mKey;
+
+  /** The retry policy of this multipart upload. */
+  protected final RetryPolicy mRetryPolicy = new CountingRetry(5);
+
+  /** Pre-allocated byte buffer for writing single characters. */
+  protected final byte[] mSingleCharWrite = new byte[1];
+
+  /** The MD5 hash of the file. */
+  protected MessageDigest mHash;
+
+  /** Flag to indicate this stream has been closed, to ensure close is only done once. */
+  protected boolean mClosed = false;
+
+  /** When the offset reaches the partition size, we upload the temp file. */
+  protected long mPartitionOffset;
+  /** The maximum allowed size of a partition. */
+  protected final long mPartitionSize;
+
+  /**
+   * The local temp file that will be uploaded when reaches the partition size
+   * or when flush() is called and this file is bigger than {@link #UPLOAD_THRESHOLD}.
+   */
+  protected File mFile;
+  /** The output stream to the local temp file. */
+  protected OutputStream mLocalOutputStream;
+
+  /**
+   * Give each upload request a unique and continuous id
+   * so that object storage knows the part sequence to concatenate the parts to a single object.
+   */
+  private final AtomicInteger mPartNumber;
+
+  /** Executing the upload tasks. */
+  private final ListeningExecutorService mExecutor;
+
+  /** Store the future of tags. */
+  private final List<ListenableFuture<?>> mFutures = new ArrayList<>();
+
+  /** upload part timeout, null means no timeout. */
+  @Nullable
+  private Duration mUploadPartTimeout;
```

Comment:
I think using nullable is enough here, the IDE will remind us to check if it is null. optional is suitable for more complex scenarios and also causes a non-negligible performance overhead.

Reply:
Hi @qian0817 , why optional introduce non-negligible performance overhead? Can you help point me to some related information? Would want to learn more into it. Thanks 
We are internally forcing the code standard to use Optional because not everybody looks into IDE advice carefully and want to give clear information to caller and future code maintainer whether the value can or cannot be zero.

Reply:
Optional is a normal generic class in Java, and it's is not zero-cost. Even though a single Optional object may make a small difference, if you use an Optional in any places, you will undoubtedly create a large number of objects and may still causes a non-negligible performance overhead.
If it is a simple judgment, for example:
use
```java
if (a != null)
```
rather than
```java
if (a.isPersent())
```
If it's a more complex scenario, for example:
use
```java
Optional.ofNullable(a).map(A::getB).map(B::getC).map(C::getD).orElse(null);
```
rather than
```java
if (a != null)
  if (a.b != null)
    if (a.b.c != null)
```